File: Apps/BankBusiness/backend/service/BusinessClientService.cs
using backend.dto.BusinessClient;
using backend.http.responses.BusinessClient;
using backend.kafka;
using backend.kafka.requests;
using backend.model;
using backend.repository;

namespace backend.service;

public class BusinessClientService(
    BusinessClientRepository repository,
    KafkaProducer producer,
    SecurityService securityService)
{
    public BusinessClientReadDto GetDtoByEmail(string email)
    {
        var client = repository.Find(email).Result;
        return new BusinessClientReadDto(
            client.OfficialName,
            client.Brand
        );
    }

    public async Task Registration(BusinessClientCreateDto dto)
    {
        await producer.produce("business-client_registration_topic",
                new BusinessClientCreateRequest(
                    dto.OfficialName,
                    dto.Brand,
                    dto.Email,
                    securityService.HashPassword(dto.Password))
            );
    }

    public long GetIdByEmail(string email)
    {
        return repository.FindIdByEmail(email).Result;
    }
}

File: Apps/BankBusiness/backend/src/main/scala/service/BusinessClientService.scala
package service

import dao.repository.{BusinessClientRepository, BusinessClientRepositoryLive}
import dto.businessclient.BusinessClientReadDto
import errors.NoEntityPresented
import model.BusinessClient
import zio.{RIO, RLayer, Task, URLayer, ZIO, ZLayer}

case class BusinessClientService(private val businessClientRepository: BusinessClientRepository) {

  def getById(id: Long): RIO[BusinessClientRepository, BusinessClient] = {
    businessClientRepository.findById(id)
      .someOrFail(NoEntityPresented())
  }

  def getByEmail(email: String): Task[BusinessClient] = {
    businessClientRepository.findByEmail(email)
      .someOrFail(NoEntityPresented())
  }

  def getIdByEmail(email: String): RIO[BusinessClientRepository, Long] = {
    businessClientRepository.findIdByEmail(email)
      .someOrFail(NoEntityPresented())
  }

  def getDtoByEmail(email: String): RIO[BusinessClientRepository, BusinessClientReadDto] = {
    businessClientRepository.findDtoByEmail(email)
      .someOrFail(NoEntityPresented())
  }

}

object BusinessClientService {
  val layer: RLayer[BusinessClientRepository, BusinessClientService] = ZLayer.fromFunction(BusinessClientService.apply _)
}
